File: src/lib/artifact.ts
import { invoke } from "@tauri-apps/api/core";
import type {
  PackageFormat,
  PackageResult,
  SignMethod,
  SignResult,
} from "@/generated/tauri-contracts";

export type { PackageFormat, PackageResult, SignMethod, SignResult };

export async function packageArtifact(params: {
  inputDir: string;
  outputPath: string;
  format?: PackageFormat;
  includeRootDir?: boolean;
}): Promise<PackageResult> {
  const { inputDir, outputPath, format = "zip", includeRootDir = true } = params;

  return await invoke<PackageResult>("package_artifact", {
    inputDir,
    outputPath,
    format,
    includeRootDir,
  });
}

export async function signArtifact(params: {
  artifactPath: string;
  method?: SignMethod;
  outputPath?: string;
  keyId?: string;
}): Promise<SignResult> {
  const { artifactPath, method = "gpg_detached", outputPath, keyId } = params;

  return await invoke<SignResult>("sign_artifact", {
    artifactPath,
    method,
    outputPath,
    keyId,
  });
}

File: src-tauri/src/commands/artifact.rs
use crate::artifact::{PackageFormat, PackageResult, SignMethod, SignResult};
use std::path::Path;

/// Package an output directory into a single artifact file.
#[tauri::command]
pub async fn package_artifact(
    input_dir: String,
    output_path: String,
    format: Option<PackageFormat>,
    include_root_dir: Option<bool>,
) -> Result<PackageResult, crate::errors::AppError> {
    let format = format.unwrap_or(PackageFormat::Zip);
    let include_root_dir = include_root_dir.unwrap_or(true);
    crate::artifact::package_directory(
        Path::new(&input_dir),
        Path::new(&output_path),
        format,
        include_root_dir,
    )
    .await
    .map_err(|e| crate::errors::AppError::unknown(format!("package failed: {}", e)))
}

/// Sign an artifact file using a supported signing method.
#[tauri::command]
pub async fn sign_artifact(
    artifact_path: String,
    method: SignMethod,
    output_path: Option<String>,
    key_id: Option<String>,
) -> Result<SignResult, crate::errors::AppError> {
    crate::artifact::sign_artifact(
        Path::new(&artifact_path),
        method,
        output_path.as_deref().map(Path::new),
        key_id.as_deref(),
    )
    .await
    .map_err(|e| crate::errors::AppError::unknown(format!("sign failed: {}", e)))
}
